Pal/src/host/Linux-SGX/db_mutex.c

```diff
@@ -123,7 +123,7 @@ int _DkMutexUnlock (struct mutex_handle * m)
     int need_wake;
 
     /* Unlock */
-    *(m->locked) = 0;
+    *(m->locked) = MUTEX_UNLOCKED; // TODO: this is not atomic!
```

Comment:
A store is atomic and I believe this is correct as used.  It just isn't an atomic compare-and-swap; as long as we are sure we are transitioning from locked to unlocked, this idiom should be correct, no?

Reply:
No, compilers don't provide such strong guarantees. In practice it will *probably* produce an atomic store on x64, but it's *not* guaranteed to do so. On the other hand, this code is not atomic on x86, because `locked` is `volatile int64_t `.

Basically, the compiler is free to split this write into two smaller ones or merge it with some adjacent writes. There's also another, more likely to happen in real life problem: volatile doesn't imply a memory barrier. So, another thread may see that `m->locked` write before seeing the last write to some other, guarded by this lock memory.

There's a reason for C to have `_Atomic` type specifier ;)

Reply:
Good points.  I retract my claim.